### ISSUE TITLE
moving data off the lua stack entirely and into C++?

### ISSUE THREAD

**Issue by TheMaverickProgrammer** [ThePhD/sol2#1122]
I only see references, objects, and the instruction to used shared pointer types in the docs so this question might be quite niche.

Essentially, I want to use scripting to return a UDT from lua and release ownership from the data created there. In my codebase, I have a very specific pipeline that ends with a pointer being de allocated a specific way. (think of a conveyor belt with no return!)

so something like

```cpp
AbstractDevice* ScriptedDevice::OnExecute() {
    std::object obj = luaState["on_execute_impl"]();
    return obj->release(); // give me the pointer! (NOTE: program WILL delete when done!
}
```

and in the lua file

```lua
local function on_execute_impl
   return Engine.Devices.ExampleDevice.new() -- assume these namespaces and UDT are properly binded in my code
end
```

I'd rather have the ability to do this than refactor my code but I understand if that's what I'll have to do if this isn't possible.

**Comment by TheMaverickProgrammer** [comment]
I see that factories might be able to produce what I want. If I return a unique_ptr from the factory method, I could `release()` it later.
But how can I grab the unique ptr out of the lua state? 

This fails for obvious reasons:

```cpp
auto ptr = obj.as<std::unique_ptr<CardAction>>(); // no copies allowed...
AbstractDevice* device = ptr->release(); // what I want!
```

**Comment by ThePhD** [comment]
```cpp
sol::object obj = luaState["on_execute_impl"]();
if (obj.valid()) {
     // notice the reference
     auto& ptr = obj.as<std::unique_ptr<CardAction>>();
     return ptr->release();
}
return nullptr;
```
----------------------
```cpp
// sol's optional, not the standard's
sol::optional<
     std::unique_ptr<AbstractDevice>& // Note the reference marker here
> maybe_device = luaState["on_execute_impl"]();
if (maybe_device.has_value()) {
     return maybe_device->release();
}
return nullptr;
```

**Comment by TheMaverickProgrammer** [comment]
```cpp
auto& ptr = obj.as<std::unique_ptr<CardAction>();
```

still fails

`attempting to reference a deleted function` sol.hpp line 11800.

The second one:

```
U must be an lvalue
a value of type "sol::protected_function_result*" cannot be used to initialize an entity of type "std::unique_ptr<CardAction, std::default_delete<CardAction>>*"

'sol::optional<std::unique_ptr<CardAction, std::default_delete<CardAction>>&>::m_value' member could not be initialized
'initializing': cannot convert from '_Ty*' to 'T*'
```

in sol.hpp lines 5528-5529

**Comment by ThePhD** [comment]
Ah. Some funky interactions are stopping the 2nd one.

2 things:

- Pull from the latest, I fixed a bug
- Use the 1st example


Then you should be good to go.

**Comment by TheMaverickProgrammer** [comment]
are you the maintainer of the vcpkg of sol? That's what I was using.

**Comment by ThePhD** [comment]
Oh, no. I guess I'd have to send an update over there...

**Comment by TheMaverickProgrammer** [comment]
unfortunately I'm getting the same error. 
I made sure I am now using the code from the latest `commit b77f1a21020bc8a090081c69481b3fbb52c6b513 `
If it helps, I am building with C++17 on VS 2019 (Windows 10)

I have done the following:

```cpp
  sol::object obj = script["execute_special_impl"](this);

  if (obj.valid()) {
  // as shown from https://github.com/ThePhD/sol2/commit/b77f1a21020bc8a090081c69481b3fbb52c6b513#diff-f31dfd6c1e73d3d45e20df66ecd2fffa7c093179004f631a023afda964f2fa26R723
    auto& ptr = obj.as<std::unique_ptr<CardAction>&>();
    return ptr.release();
  }
  ```
1. removed sol2 from my cmake. did a clear cache.
2. cloned sol2. Copied the single/include/sol header files and put them directly into my project source directory.
3. change my code as shown above

Error: 

`std::unique_ptr<CardAction, std::default_delete<CardAction>>::unique_ptr(const std::unique_ptr<CardAction, std::default_delete<CardAction>>&): attempting to reference a deleted function`

So to be sure, I also did another round:
1. clear cmake cache, 
2. ran `single.py` from the sol directory to ensure the contents of the header files were up to date!
3. complete project rebuild

And have the same error.
